when kubelet restarted, keep the readiness value intact until probed again

langyenan opened the pull request:
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
if the kubelet is restart, the status of the pod in the node will be set to failure, and then set to ready. the pod will be remove from ep and than add to ep. If the number of the service is 1, the result is that the service can't connect by other service at that duration, but the pod is work. if lots of kubelet is restart, it will affect more services

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #78733

**Special notes for your reviewer**:
tested on kubernetes v1.16.2

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

mattjmcnaughton commented:
cc @zouyee 
xref #79555

@langyenan and @zouyee do you mind syncing on y'alls diffs, which I believe are trying to solve the same problem. It would be great to have us all on the same page if the solutions proposed in each diff are different, and if so, what are the pros and cons of each approach.

zouyee commented:
@mattjmcnaughton thanks，I have no problem with that.
/cc @derekwaynecarr

langyenan commented:
@mattjmcnaughton  thanks, I'm totally ok with that, would you tell me how should I sync the diffs with #79555 submitted by @zouyee  ？Actually, the midoficatin in  #79555 is included in this pr

langyenan commented:
/assign @derekwaynecarr

langyenan commented:
@zouyee Hi, is there anything I can do to get this PR reviewed?

mattjmcnaughton commented:
ping @zouyee - mind coordinating with @langyenan about the best path forward? thanks!

langyenan commented:
This PR failed some ut, I'll push another commit when I correct it

langyenan commented:
/cc @zouyee @mattjmcnaughton Hi, I fixed and added some unit tests, PTAL, thanks!

Here's what I did in the 2 commits:
In db9a523128d926d98ac6db31c079210fa7f44ab2, following changes are made:

1. In the `doProbe()`, I copied the code from @zouyee 's commit(#79555), it prevents the worker from initializing the readiness state before running the probe

2.  `probeManager.UpdatePodStatus()` will use the readiness state from api-server if it doesn't have any result reported by worker. This is necessary because `UpdatePodStatus()` is invoked earlier than the worker

In bde2f0a41773bf97adc2340069432b9c0381cd0c, I fixed 3 unit tests and added 2 test cases:

1. 3 fixed unit tests: `TestResultRunOnStartupCheckFailure`, `TestLivenessProbeDisabledByStarted` and `TestStartupProbeDisabledByStarted`.  The thought behind the changes is similar: a worker sets the result if and only if corresponding threshold is reached.

2. 1 new test case in `TestUpdatePodStatus`:  a container named `unprobedReady` is added, it's readiness state is `true` (updated before kubelet restarted), it's state should not be changed by `UpdatePodStatus()`.

3. 1 new unit test `TestReadinessProbeOnKubeletRestart`:  this is a simulation of kubelet restarting.

hex108 commented:
/priority important-soon
/ok-to-test

hex108 commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -1352,7 +1352,13 @@ func (kl *Kubelet) generateAPIPodStatus(pod *v1.Pod, podStatus *kubecontainer.Po
+			oldContainerReadiness[c.ContainerID] = c.Ready
+		}
+	}
+	kl.probeManager.UpdatePodStatus(pod.UID, s, oldContainerReadiness)
```
Could you update `s.Status.ContainerStatuses` to `oldContainerReadiness` directly? Then we do not need to add another argument `oldContainerReadiness`.

langyenan replied:
Yes, it's simpler that way, let me handle it, thanks!

hex108 commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -1352,7 +1352,13 @@ func (kl *Kubelet) generateAPIPodStatus(pod *v1.Pod, podStatus *kubecontainer.Po
 	}
-	kl.probeManager.UpdatePodStatus(pod.UID, s)
+	oldContainerReadiness := make(map[string]bool)
+	if p, found := kl.podManager.GetPodByName(pod.Namespace, pod.Name); found{
```
It might be better to add a comment that these old container readiness status are from apiserver and explain why we need to use them as pod container readiness status.

langyenan replied:
Sure, thanks!

zouyee commented:
BTW, you need format your commit.
`./hack/update-gofmt.sh`

langyenan commented:
> BTW, you need format your commit.
> `./hack/update-gofmt.sh`

My bad, thanks

langyenan committed:
when kubelet restarted, keep the readiness value intact until probed again

langyenan committed:
unit tests for container readiness modification

langyenan commented:
/test pull-kubernetes-bazel-test

hex108 commented on pkg/kubelet/prober/prober_manager.go:
```diff
@@ -241,6 +241,9 @@ func (m *manager) UpdatePodStatus(podUID types.UID, podStatus *v1.PodStatus) {
 		} else if result, ok := m.readinessManager.Get(kubecontainer.ParseContainerID(c.ContainerID)); ok {
 			ready = result == results.Success
+		} else if c.Ready {
+			// keep the readiness until we get a new result from readinessManager
```
It might be better if adding what cases you are dealing with.

langyenan replied:
Ok, I'll add the issue number in the comment

langyenan commented on pkg/kubelet/prober/worker.go:
```diff
@@ -201,7 +201,9 @@ func (w *worker) doProbe() (keepGoing bool) {
 		}
 		w.containerID = kubecontainer.ParseContainerID(c.ContainerID)
-		w.resultsManager.Set(w.containerID, w.initialValue, w.pod)
+		if c.State.Running == nil || int32(time.Since(c.State.Running.StartedAt.Time).Seconds()) < w.spec.InitialDelaySeconds {
```
Without it, the readiness state is always set `false`, even if the container was ready before kubelet restarted

langyenan commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -1458,6 +1458,9 @@ func (kl *Kubelet) convertToAPIContainerStatuses(pod *v1.Pod, podStatus *kubecon
 		switch cs.State {
 		case kubecontainer.ContainerStateRunning:
 			status.State.Running = &v1.ContainerStateRunning{StartedAt: metav1.NewTime(cs.StartedAt)}
+			if pc, found := podutil.GetContainerStatus(previousStatus, cs.Name); found && pc.ContainerID == cid {
```
Ok, I'll add some comment here

hex108 reviewed (commented):
/lgtm

Also ping @yujuhong for approval.

langyenan commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu
/test pull-kubernetes-e2e-gce

langyenan commented:
/assign @Random-Liu
/assign @tallclair

tallclair commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -1458,6 +1458,11 @@ func (kl *Kubelet) convertToAPIContainerStatuses(pod *v1.Pod, podStatus *kubecon
 		switch cs.State {
 		case kubecontainer.ContainerStateRunning:
 			status.State.Running = &v1.ContainerStateRunning{StartedAt: metav1.NewTime(cs.StartedAt)}
+			// Use the ready value from previousStatus so that we don't initialize it to false
```
I'm confused by this. Isn't this change just delaying setting the status for 1 cycle in all cases? I don't think we want that.

langyenan replied:
Since kubelet always initialize status.Ready to false, all `ready` pods are marked `not ready` after kubelet restart, those pods are then kicked from Endpoints. This code keep the value of status.Ready received from api-server so that it will not be set to `false` until probe worker finishes the test.

tedyu commented on pkg/kubelet/prober/worker_test.go:
```diff
@@ -464,3 +476,21 @@ func TestStartupProbeDisabledByStarted(t *testing.T) {
+	// given a ready container with FailureThreshold set 2, following condition should met:
+	// 1. probe worker should not initialize the readiness to false
+	// 2. the 1st probe result should not update to resultManager (because the threshold is not reached)
+	// PS: this is a simulation of kubelet restart
```
Can you add more comment on how this test simulates the kubelet restart ?
